### ISSUE TITLE
Race condition in creation of AutoML temp folders

### ISSUE THREAD

**Issue by justinormont** [dotnet/machinelearning#3803]
**tldr;** recommend a GUID based folder name: `Temp\Microsoft.ML.AutoML\experiment_9bdaa79e-8ceb-4103-988b-9d73aefd53c2\...`

User reported bug by @markusmobius in https://github.com/dotnet/machinelearning/issues/3749

## Bug
AutoML stores the trained models from the sweeping process on disk to avoid ever growing RAM usage. We currently ensure the folder is named uniquely by checking if it exists, and if so, incrementing a counter at the end. This process has a race condition where multiple processes will test for the folder's existence, see nothing, and both create and begin to work in the same folder. 

## Background
@markusmobius was running multiple instances of the AutoML CLI, and the various processes were clobbering each others' temporary folders. See https://github.com/dotnet/machinelearning/issues/3749#issuecomment-495259293

As said by @daholste in https://github.com/dotnet/machinelearning/issues/3749#issuecomment-495407319:
> Re: writing models to the temp folder -- when initializing experiments one after the other on a machine, each experiment would have its own unique model folder:
> ![image](https://user-images.githubusercontent.com/43974253/58290281-0c6f3500-7d6e-11e9-8790-d8a4077dab21.png)
> However, when initializing many experiments in parallel on the same machine, it is definitely possible to have a race condition that causes conflict here. Thanks for pointing this out 

Reply by @justinormont in https://github.com/dotnet/machinelearning/issues/3749#issuecomment-495426783 / https://github.com/dotnet/machinelearning/issues/3749#issuecomment-495448847: 
> I believe we used a named mutex in the word embedding's model downloader to ensure only one copy of the model is being downloaded per machine. 

> I found the word embedding's named mutex: https://github.com/dotnet/machinelearning/blob/b861b5d64841cbe0f2c866ee7586872aac450a51/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L237-L246
> 
> The lock is in the downloader class (resource manager).
> 
> Plus the `finally` unlock:
> https://github.com/dotnet/machinelearning/blob/b861b5d64841cbe0f2c866ee7586872aac450a51/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L287-L291
> 
> We could do the same for the temp folder; if I recall I thought it was race free when the storage is local, but not for network shares due to lack of a remote atomic rename. 
> 
> More simple approach is to use a GUID: `Temp\Microsoft.ML.AutoML\experiment_9bdaa79e-8ceb-4103-988b-9d73aefd53c2\...`

**Comment by jwood803** [comment]
@justinormont It looks like the path comes from the `GetFilePath` method in that class. 

https://github.com/dotnet/machinelearning/blob/b861b5d64841cbe0f2c866ee7586872aac450a51/src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs#L191-L197

Would the same fix for #568 apply here as well, like below?

```
var filePath = Path.Combine(absDir, Guid.NewGuid().ToString(), fileName);
```

**Comment by justinormont** [comment]
@jwood803: Yes, the same style of fix would apply.

I think the cleanest way of fixing is to replace:
https://github.com/dotnet/machinelearning/blob/0435f6b3ae976967b532e164df50e71e56a22d07/src/Microsoft.ML.AutoML/Experiment/Experiment.cs#L115-L127

I expect the replacement would be:
```c#
var experimentDirFullPath = Path.Combine(rootDir.FullName, "experiment_" + Guid.NewGuid().ToString()); 
```

This creates a path of `Temp\Microsoft.ML.AutoML\experiment_{GUID}\...`. 

The GUID would be stable for the full AutoML sweep (experiment), so all of the models for a single AutoML sweep are grouped together. And when we delete the temporary models for a sweep, it's easy to just clear out this whole folder (to fix https://github.com/dotnet/machinelearning-modelbuilder/issues/320).

**Comment by justinormont** [comment]
Closed by https://github.com/dotnet/machinelearning/pull/4657
